### DISCUSSION TITLE
How to reduce memory usage while creating variables and constraints?

### DISCUSSION THREAD

**Discussion by sarthakgaur** [google/or-tools#2778] [CP-SAT questions] [answered]
We have a problem similar to the [employee scheduling](https://developers.google.com/optimization/scheduling/employee_scheduling) problem. The amount of variables that we need to create are quite large, and we have a lot of constraints. The program is written in Python and it works fine for a small data set. But, for large data sets the memory usage gets very high. It got to a point that I had to stop the program because I didn't have any memory left in my 8 GB system. This happened before the solver even started. 

This program will run in some server in future, and we will have even larger data sets. So, we'll have to do our best to reduce the memory usage.

Right now, the program is written in plain Python. We are are not using libraries like NumPy. The other option is to write the program in C++.

Using a library like NumPy can reduce the memory usage. And, there are some changes that we can do to reduce the amount of variables that we are creating. But, maybe writing the program in C++ will be better.

I need some suggestions on this. Which approach do you think will work for us? Any help will be appreciated! Thanks a lot!

**Comment by lperron** [comment, marked as answer]
A good start would to be to count how many python objects you are creating.

**Reply by sarthakgaur** [reply]
We were creating a lot of boolean variables and then adding constraints on them to set them to 0. We removed all these boolean variables. This also reduced the amount of constraints we have per variable. The memory usage is good for now.

Thanks for replying!
